### ISSUE TITLE
Error in processing Fortran file with character string returning function

### ISSUE THREAD

**Issue by Slookeur** [doxygen/doxygen#10726]
**Describe the bug**

Doxygen is not able to process Fortran files that include character string returning function:
For example:

```
CHARACTER (LEN=16) FUNCTION this_example ()

this_example = "This is a test !"

END FUNCTION

SUBROUTINE test_this ()

IMPLICIT NONE

INTERFACE
  CHARACTER(LEN=16) FUNCTION this_example ()
  END FUNCTION
END INTERFACE

write (6, *) this_example ()

END SUBROUTINE
```

The error message when running Doxygen is the following: 

**line: (last line of the file): state: 22(String)**

Thank you for developing Doxygen !
And thanks in advance for the time you will take to consider this.

**Comment by albert-github** [comment]
When I place the code in a file **a.f90**, use a default settings file and run with the current doxygen version (1.10.0) I get:
```
aa.f90 line: 19, state: 10(ModuleBody)
```
which is different from your message (but still not OK)

So:
- Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file!) that allows us to reproduce the problem? Please don't add external links as they might not be persistent (also references to GitHub repositories are considered non persistent).
- Please also specify the **full** doxygen version used (`doxygen -v`).

**Comment by Slookeur** [comment]
Hello and thank you for your quick answer, 
first of all you can find the file that really gave me the issue here: 

https://github.com/Slookeur/Atomes

If you clone the repo, check `src/fortran/writedata.F90`, I used `!!\cond` and `!!\endcond` comments to stop Doxygen from processing the mentioned `FUNCTION` instructions in this file, simply remove the comments, then run a basic Doxygen on the `src` directory to have more information. (you can use the files in `atomes-doxygen` but you will need to edit the files to correct the file paths for your needs.)

In the mean time I will work on a basic sample code for you that reproduces the error. 

Regards.

**Comment by albert-github** [comment]
For the problem I saw:
- I've just pushed a proposed patch, pull request #10727


I quickly tested the writedata.F90 file and with the old doxygen version I got the warning as indicated ()when removing the `\cond` and `\endcond`, with the updated version it looks all OK to me.

In the source code I see a small problem e.g. line 653 in the source code reads:
```
      ylegend= "Correction (z) [Å²]"
```
though doxygen in its source browser think it is:
![image](https://github.com/doxygen/doxygen/assets/5223533/5b1f5b86-438f-406f-b35f-a6141e3f8a71)

something for a separate  issue / fix.

**Comment by Slookeur** [comment]
You're fast ;-)

Thank you for all this information, when I tried to isolate the issue I notice that I needed to comment the entire 2 first FUNCTIONS of the file `CHARACTER (LEN=35) FUNCTION ylegend (job, nleg, idl)` and `CHARACTER (LEN=65) FUNCTION xlegend (job, nleg, idl, cdc)` to get Doxygen to work properly, but considering the number of strings and peculiar characters in there I was not sure that it might not be an issue as well. 

So far no way to produce a small code with the issue ... I tried with both Doxygen 1.9.6 and 1.10.0

I will keep looking for the real problem, when ready I will post a new issue with a proper and simple example. 

Thank you for the time you took to look into this, and again thanks for Doxygen ;-)

**Comment by albert-github** [comment]
> I will keep looking for the real problem, when ready I will post a new issue with a proper and simple example.

The real problem is that doxygen heuristically tries to determine whether or not we have fixed or free Fortran formatted code.
As we have here in the first position a `C` doxygen thought it was fixed formatted code whilst in fact it is free formatted code, I now added some "exception" words.

I think interesting for you might also be the setting `EXTENSION_MAPPING` where one could set `EXTENSION_MAPPING = F90=FortranFree`

**Comment by albert-github** [comment]
Code of both PR has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.11.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
